generators/python-v2/base/src/context/AbstractPythonGeneratorContext.ts

```diff
@@ -0,0 +1,57 @@
+import { AbstractGeneratorContext, FernGeneratorExec, GeneratorNotificationService } from "@fern-api/generator-commons";
+import { IntermediateRepresentation, Name, TypeDeclaration, TypeId, TypeReference } from "@fern-fern/ir-sdk/api";
+import { snakeCase } from "lodash-es";
+import { BasePythonCustomConfigSchema } from "../custom-config/BasePythonCustomConfigSchema";
+import { PythonProject } from "../project";
+import { PythonTypeMapper } from "./PythonTypeMapper";
+
+export abstract class AbstractPythonGeneratorContext<
+    CustomConfig extends BasePythonCustomConfigSchema
+> extends AbstractGeneratorContext {
+    private packageName: string;
+    public readonly pythonTypeMapper: PythonTypeMapper;
+    public readonly project: PythonProject;
+
+    public constructor(
+        public readonly ir: IntermediateRepresentation,
+        public readonly config: FernGeneratorExec.config.GeneratorConfig,
+        public readonly customConfig: CustomConfig,
+        public readonly generatorNotificationService: GeneratorNotificationService
+    ) {
+        super(config, generatorNotificationService);
+        this.packageName = snakeCase(`${this.config.organization}_${this.ir.apiName.snakeCase.unsafeName}`);
+        this.pythonTypeMapper = new PythonTypeMapper(this);
+        this.project = new PythonProject({ context: this });
+    }
+
+    public getTypeDeclarationOrThrow(typeId: TypeId): TypeDeclaration {
+        const typeDeclaration = this.ir.types[typeId];
+        if (typeDeclaration == null) {
+            throw new Error(`Could not find type declaration for type id: ${typeId}`);
+        }
+        return typeDeclaration;
+    }
+
+    public isTypeReferenceOptional(typeReference: TypeReference): boolean {
+        if (typeReference.type === "container" && typeReference.container.type === "optional") {
+            return true;
+        }
+        if (typeReference.type === "named") {
+            const typeDeclaration = this.getTypeDeclarationOrThrow(typeReference.typeId);
+            if (typeDeclaration.shape.type === "alias") {
+                return this.isTypeReferenceOptional(typeDeclaration.shape.aliasOf);
+            }
+        }
+        return false;
+    }
+
+    public getPascalCaseSafeName(name: Name): string {
+        return name.pascalCase.safeName;
```

Comment:
Should we be using `unsafeName` here since they shouldn't ever conflict with Python keywords (e.g. we can use `Str` instead of `Str_`)? Does this preserve behavior with the current generator?

Reply:
Yup copied the existing generator for now, and was hoping to tackle as i went deeper to see what could be improved (but also not create compile breaks)